### PR TITLE
Fix building docker image proccess

### PR DESCRIPTION
**Pull request by dkropachev**
https://trello.com/c/qmS9onLT/1595-fix-issues-with-docker-backend
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

**Comment by dkropachev** [comment]
Localy tested - worked whell, but provision tests are failed, to be addressed in another PR

File: vars/getJenkinsLabels.groovy [hunk 1]
```diff
@@ -12,7 +12,8 @@ def call(String backend, String aws_region=null) {
 
     def jenkins_labels = ['aws-eu-west-1': 'aws-sct-builders-eu-west-1',
                           'aws-us-east-1' : 'aws-sct-builders-us-east-1',
-                          'gce': 'gce-sct-builders']
+                          'gce': 'gce-sct-builders',
+                          'docker': 'gce-sct-builders']
```
Comment (fruch) [inline vars/getJenkinsLabels.groovy]: Why docker can't run on all `sct-builders`

File: sdcm/docker.py [hunk 2]
```diff
@@ -157,7 +162,7 @@ def _update_image(self):
 
     def _create_container(self, node_name, is_seed=False, seed_ip=None):
         labels = f"--label 'test_id={cluster.Setup.test_id()}'"
-        cmd = f'run --cpus="1" --name {node_name} {labels} -d {self._node_img_tag}'
+        cmd = f'run --privileged --cpus="1" --name {node_name} {labels} -d {self._node_img_tag}'
```
Comment (fruch) [inline sdcm/docker.py]: Why needs to run privileged ?

File: sdcm/docker.py [hunk 3]
```diff
@@ -141,8 +141,13 @@ def __init__(self, **kwargs):
 
     def _create_node_image(self):
         self._update_image()
+        private_key = self.credentials[0].key_file
```
Comment (fruch) [inline sdcm/docker.py]: Why this is not done as part of the image Dockerfile, why do we need to create this here ?

File: docker/scylla-sct/Dockerfile [hunk 4]
```diff
@@ -30,6 +30,7 @@ ADD etc/collectd.conf /etc/supervisord.conf.d/collectd.conf
 ADD etc/scylla-manager.conf /etc/supervisord.conf.d/scylla-manager.conf
 RUN echo "autostart=false" >> /etc/supervisord.conf.d/scylla-server.conf
 RUN echo "autostart=false" >> /etc/supervisord.conf.d/scylla-housekeeping.conf
+RUN echo "sysctl --system" >> /etc/rc.local
```
Comment (fruch) [inline docker/scylla-sct/Dockerfile]: What's this is needed for ?

File: docker/scylla-sct/etc/sysctl.conf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+[program:collectd]
+command=/usr/sbin/sysctl --system
```
Comment (fruch) [inline docker/scylla-sct/etc/sysctl.conf]: Again, why we need this or process ? For what exactly ?

**Comment by bentsi** [comment]
@dkropachev can you answer Israel's questions?

**Comment by dkropachev** [comment]
@fruch,
 If you don't run sysctl --system you will end up with following issue on db nodes:

[root@68e140b2745f /]# scylla
Scylla version 3.2.0-0.20200115.f9b11c9b30 starting ...
command used: "/usr/bin/scylla"
parsed command line options: []
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: WARNING: unable to mbind shard memory; performance may suffer: 

WARNING: unable to mbind shard memory; performance may suffer: 
ERROR 2020-02-03 11:47:10,936 [shard 0] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application

**Comment by fruch** [comment]
@dkropachev 

```
ERROR 2020-02-03 11:47:10,936 [shard 0] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
```
on my machine:
```bash
$ cat /proc/sys/fs/aio-max-nr
1048576
```

you have lower then that ?

File: vars/getJenkinsLabels.groovy (hunk 1 above)
Comment (fruch) [reply vars/getJenkinsLabels.groovy]: but less slots to run it.... all those machines are up and running and we pay for all of them :)

File: sdcm/docker.py (hunk 2 above)
Comment (fruch) [reply sdcm/docker.py]: I'm running it locally, and with hydra.
and didn't need the `systctl --system` nor the `--privileged`

I need more details why this is needed, cause I see nor reason it should.
in your log you called scylla directly without the correct commandline args.
please use `supervisorctl start scylla` and verify those extra processes and permissions are actually needed

Comment (dkropachev) [reply sdcm/docker.py]: > I'm running it locally, and with hydra.
> and didn't need the `systctl --system` nor the `--privileged`
> 
> I need more details why this is needed, cause I see nor reason it should.
> in your log you called scylla directly without the correct commandline args.
> please use `supervisorctl start scylla` and verify those extra processes and permissions are actually needed

You have it working because your host system sysctl variables was tweeked, probably you have installed scylla on it at some point in the past. 
Docker does not support sysctl modifications from non-privileged containers.

Here is what I have:
[root@some-scylla /]# supervisorctl start scylla
scylla: ERROR (abnormal termination)
[root@some-scylla /]# supervisorctl
node-exporter                    RUNNING   pid 30, uptime 0:04:45
scylla                           FATAL     Exited too quickly (process log may have details)

[root@some-scylla /]# scylla --options-file /etc/scylla/scylla.yaml 
Scylla version 3.3.rc0-0.20200204.00b3f281995 with build-id 896009254492dd09d3b3094b5fc4d4bf559486e7 starting ...
command used: "/usr/bin/scylla --options-file /etc/scylla/scylla.yaml"
parsed command line options: [options-file: /etc/scylla/scylla.yaml]
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
ERROR 2020-02-06 17:54:20,155 [shard 3] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application

Comment (fruch) [reply sdcm/docker.py]: I'm o.k. with documenting what we need todo with the host to run the docker image.
and then we do that once on where we need.

Comment (dkropachev) [reply sdcm/docker.py]: > I'm o.k. with documenting what we need todo with the host to run the docker image.
> and then we do that once on where we need.

I'am sorry, did not get that. What exactly should I do ?

Comment (fruch) [reply sdcm/docker.py]: I don't know exactly, since it was working for me out of the box. (And yes, I did install scylla a few time locally, maybe it changed something, maybe I did manually months ago for other reasons)

let start with a higher value in `/proc/sys/fs/aio-max-nr` and work it out from there, understanding what's really missing.

Comment (dkropachev) [reply sdcm/docker.py]: > I don't know exactly, since it was working for me out of the box. (And yes, I did install scylla a few time locally, maybe it changed something, maybe I did manually months ago for other reasons)
> 
> let start with a higher value in `/proc/sys/fs/aio-max-nr` and work it out from there, understanding what's really missing.

Scylla installer already does everything for us, only thing we need to do is to run sysctl --system to pick it up from inside of container.

File: vars/getJenkinsLabels.groovy (hunk 1 above)
Comment (bentsi) [reply vars/getJenkinsLabels.groovy]: @dkropachev I don't understand how are they cheaper. The builders are on-demand instances running all the time. Please change to `sct-builders` as Israel recommended

**Comment by bentsi** [comment]
@fruch AFAIK we still don't have provision test for the Docker backend. Right?
(only label)
image

**Comment by fruch** [comment]
@bentsi, yes. I've seen it, and tried using it... before looking at the PR. :disappointed: 

anyhow @bentsi, seem like me and @dkropachev need you call on using `sysctl --system` and `--privileged` (I see you marked it resolved, but didn't understood what is the resolution)

**Comment by bentsi** [comment]
@fruch you probably ran Scylla locally before you ran official docker image or you configured "aio-max-nr" to something more than 65536, that's why it worked for you. On my laptop, I am using:
```
bentsi@bentsi-G3-3590:~/devel/scylladb$ cat /proc/sys/fs/aio-max-nr
65536

```
Looks like starting from 3.1 version Scylla can't start :( this is a serious regression (last working Docker is `scylla:3.0.10`).
Tried on my laptop latest (3.3.rc1-0.20200209)
`docker run --name scylla -d scylladb/scylla`
Scylla didn't start automatically. When running again it says `Could not setup Async I/O`
```
[root@4344a189365a /]# pgrep -laf scylla
1 /opt/scylladb/python3/bin/python3 /opt/scylladb/python3/bin/../libexec/python3.7.bin -s /docker-entrypoint.py
28 /opt/scylladb/jmx/symlinks/scylla-jmx -Xmx256m -XX:+UseSerialGC -XX:+HeapDumpOnOutOfMemoryError -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.host=localhost -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=7199 -Djava.rmi.server.hostname=localhost -Dcom.sun.management.jmxremote.rmi.port=7199 -Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder -jar /opt/scylladb/jmx/scylla-jmx-1.0.jar
32 /bin/bash /scylla-housekeeping-service.sh
[root@4344a189365a /]# scylla
Scylla version 3.3.rc1-0.20200209.0d0c1d43188 with build-id d6a28c1f87e959483c864fd44fe3606d484d10b6 starting ...
command used: "/usr/bin/scylla"
parsed command line options: []
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
WARNING: unable to mbind shard memory; performance may suffer: 
ERROR 2020-02-11 11:22:27,989 [shard 7] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
[root@4344a189365a /]# exit

```
@enaydanov we will need to test this situation in the Docker artifact test
CC: @roydahan

**Comment by fruch** [comment]
@bentsi 
seems like https://github.com/scylladb/scylla/issues/5638 is exactly about the issue you are having

**Comment by bentsi** [comment]
per our discussion with Israel
1. Let's run the container without `--privileged` mode
2. Israel already altered `/proc/sys/fs/aio-max-nr` on all builders, so you can only add a check in SCT that checks that it is set correctly

**Comment by bentsi** [comment]
docker provision test failed. please check why

File: Jenkinsfile [hunk 6]
```diff
@@ -31,6 +34,10 @@ def runSctProvisionTest(String backend){
 }
 
 def runCollectLogs(String backend){
+    def sct_config_files = "test-cases/PR-provision-test.yaml"
+    if (backend == 'backed') {
```
Comment (bentsi) [inline Jenkinsfile]: what does it mean `backed`

Comment (dkropachev) [reply Jenkinsfile]: fixed

File: sdcm/cluster.py [hunk 7]
```diff
@@ -4054,6 +4054,7 @@ def start_scylla_monitoring(self, node):
         run_script = dedent("""
             cd {0.monitor_install_path}
             mkdir -p {0.monitoring_data_dir}
+            ./kill-all.sh
```
Comment (bentsi) [inline sdcm/cluster.py]: why did you add this?

Comment (bentsi) [reply sdcm/cluster.py]: please remove this and merge.
We need to implement "clean resources" for Docker. I will open a task for that

**Comment by dkropachev** [comment]
This PR will work after https://github.com/scylladb/scylla-monitoring/pull/820 went it's way into scylla-monitoring

File: docker/env/hydra.sh [hunk 8]
```diff
@@ -102,4 +102,4 @@ docker run --rm ${TTY_STDIN} --privileged \
     --net=host \
     --name=${SCT_TEST_ID} \
     scylladb/hydra:${VERSION} \
-    /bin/bash -c "${TERM_SET_SIZE} eval '${CMD}'"
+    /bin/bash -c "sudo rm -f $WORK_DIR; sudo ln -s \"$SCT_DIR\" \"$WORK_DIR\"; cd \"$WORK_DIR\"; ${TERM_SET_SIZE} eval '${CMD}'"
```
Comment (bentsi) [inline docker/env/hydra.sh]: why are you removing WORK_DIR?

File: sdcm/docker.py [hunk 9]
```diff
@@ -141,27 +146,38 @@ def __init__(self, **kwargs):
...
         if images:
+            self._clean_not_user_containers()  # Remove old containers that could block removing dangling images
             _cmd(f'rmi {images}', timeout=90)
 
     def _update_image(self):
         LOGGER.debug('update scylla image')
         _cmd(f'pull {self._image}:{self._version_tag}', timeout=300)
-        self._clean_old_images()
+        try:
+            self._clean_old_images()
```
Comment (bentsi) [inline sdcm/docker.py]: `docker system prune --volumes -f`
cleans:
  - all stopped containers
  - all networks not used by at least one container
  - all dangling images
  - all dangling build cache

File: sdcm/docker.py [hunk 10]
```diff
@@ -309,29 +327,35 @@ def send_receive_files(self, src, dst, delete_dst=False, preserve_perm=True, pre
 
 
 class DummyMonitoringNode(cluster.BaseNode):  # pylint: disable=abstract-method,too-many-instance-attributes
```
Comment (bentsi) [inline sdcm/docker.py]: rename

Comment (bentsi) [reply sdcm/docker.py]: `DummyMonitoringNode` -> `DockerMonitoringNode`

File: sdcm/cluster.py [hunk 11]
```diff
@@ -4132,14 +4138,14 @@ def reconfigure_scylla_monitoring(self):
...
-            mkdir -p {0.monitoring_data_dir}
+        run_script = dedent(f"""
+            cd -P {self.monitor_install_path}
+            mkdir -p {self.monitoring_data_dir}
             ./start-all.sh \
-            -s {0.monitoring_conf_dir}/scylla_servers.yml \
-            -n {0.monitoring_conf_dir}/node_exporter_servers.yml \
-            -d {0.monitoring_data_dir} -l -v master,{0.monitoring_version} -b "-web.enable-admin-api"
-        """.format(self))
+            -s `realpath "{self.monitoring_conf_dir}/scylla_servers.yml"` \
```
Comment (bentsi) [inline sdcm/cluster.py]: @dkropachev do we still need this? if not remove this in #1852
